### ISSUE TITLE
[VMR] Build failure with portable RID (e.g. /p:TargetRid=linux-x64)

### ISSUE THREAD

**Issue by directhex** [dotnet/source-build#3131]
I'm still working on the exact root cause, but I wanted to have a written record somewhere other than internal Teams.

Building for a generic, portable RID fails currently with the VMR, in `aspnetcore`, citing that it can't find the crossgen2 binary. Looking at the build, and comparing to a "good" `ubuntu.22.04-x64` build, it seems the crossgen2 nupkg is created as expected during the runtime build, but never restored in aspnetcore.

It appears the build explicitly rejects most of the package references from earlier in the build, in fact, citing that they do not need to be explicitly depended on because they're part of the SDK.

This also blocks building an osx-x64 RID build on Mac.

**Comment by directhex** [comment]
```
   /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Runtime.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/.dotnet/sdk/7.0.100-rc.2.22477.23/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.Shared.targets(152,5): error NETSDK1023: A PackageReference for 'Microsoft.NETCore.App.Crossgen2.linux-x64' was included in your project. This package is implicitly referenced by the .NET SDK and you do not typically need to reference it from your project. For more information, see https://aka.ms/sdkimplicitrefs [/home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj]
    /home/directhex/Projects/dotnet/src/aspnetcore/artifacts/source-build/self/src/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj(434,5): error : Could not find crossgen2 tools/crossgen2

```

**Comment by directhex** [comment]
Adding `DisableImplicitFrameworkReferences` to `Microsoft.AspNetCore.App.Runtime.csproj` reduces the error count from 17 to 13, which is not very meaningful or useful :|

**Comment by omajid** [comment]
cc @tmds

**Comment by tmds** [comment]
There are some changes on aspnetcore 7.0 branch for eliminating the portable build.

The PRs against the installer repo haven't been backported yet (https://github.com/dotnet/installer/pull/14549 and https://github.com/dotnet/installer/pull/14938). Backporting those may fix this issue.

cc @crummel @MichaelSimons

**Comment by directhex** [comment]
Funny story, I think the 7.0 aspnetcore changes (specifically https://github.com/dotnet/aspnetcore/pull/44978) are the problem, not the cure.

`Dependencies.props` includes a hardcoded dependency on `Microsoft.NETCore.App.Runtime.linux-x64` et al, and _also_ includes a dependency on `Microsoft.NETCore.App.Runtime.$(TargetRuntimeIdentifier)` which (when building on Mac, or on Linux with `/p:TargetRid=linux-x64`) resolves to a duplicate dependency. This causes the `CheckForImplicitPackageReferenceOverrides` task to freak out, remove _both_ copies of the reference, and fail the build.

Locally, removing one of the duplicated dependencies progresses the build further. Without the duplicate dependency, `DisableImplicitFrameworkReferences` has no impact at all either enabled or disabled, so it turns out this was a bad area of investigation.

**Comment by directhex** [comment]
The build still fails, mind you, but later in the build.

**Comment by directhex** [comment]
System.Formats.Asn1.dll, System.Text.Encodings.Web.dll and System.Text.Json.dll exist in both the runtime pack and aspnetcore build, and crossgen2 fails if there are non-unique filenames in input. Now why does this work without the portable RID I wonder...

**Comment by directhex** [comment]
OK, no, the 2-assembly thing is a weird problem with a dirty build. Doing a 100% clean build, this is enough to make the build work:

```diff --git a/src/aspnetcore/eng/Dependencies.props b/src/aspnetcore/eng/Dependencies.props
index 6c44830a0..904a23489 100644
--- a/src/aspnetcore/eng/Dependencies.props
+++ b/src/aspnetcore/eng/Dependencies.props
@@ -100,7 +100,6 @@ and are generated based on the last package release.
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.win-arm64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.osx-x64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.osx-arm64" />
-    <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.linux-x64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.linux-arm" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.linux-arm64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Runtime.linux-musl-x64" />
@@ -117,7 +116,6 @@ and are generated based on the last package release.
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.linux-musl-arm" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.linux-musl-arm64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.linux-musl-x64" />
-    <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.linux-x64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.win-x64" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.win-x86" />
     <LatestPackageReference Include="Microsoft.NETCore.App.Crossgen2.win-arm" />

```

**Comment by directhex** [comment]
Got a Mac build. I'll post patches & file issues tomorrow morning, since changes are needed.

**Comment by directhex** [comment]
I'm going to close this. I've split the issues I encountered into separate issues.

https://github.com/dotnet/source-build/issues/3149

https://github.com/dotnet/source-build/issues/3150

https://github.com/dotnet/source-build/issues/3151

**Comment by corngood** [comment]
I'm still getting this problem on 9.0.0-preview.2 unless I make the same change described in https://github.com/dotnet/source-build/issues/3131#issuecomment-1334177681 (removing the duplicate references).

I also had a similar problem in 8.0 releases, which I described here: https://github.com/dotnet/installer/pull/15163#issuecomment-1873396096.  I should have probably posted that comment on this issue instead.

My fix from 8.0 no longer applies, but I verified that aspnetcore is building with `/p:PortableBuild=true /p:TargetRuntimeIdentifie
r=linux-x64`, so I'm not sure why it's failing again.  I'll try to find out.
